src/charm.py

```diff
@@ -1234,6 +1353,25 @@ def _run_diagnostic_command(self, cmd) -> None:
         except subprocess.CalledProcessError as e:
             logger.error(f"Exception occurred running '{cmd}'\n {e}")
 
+    def _reconfigure_replica_set(self) -> None:
+        """Optionally reconfigure replicaset if needed.
+
+        This should happen only when we are reusing storage from one cluster to another cluster,
+        because all IPs have changed and all configurations are broken.
+        """
+        if not self.db_initialised:
```

Comment:
should we let non-leaders execute this?

Reply:
Yes and we have to.
This code allows each node to fix it's local replica set configuration.
This is stored in the local database which is not shared among the replica set and this code updates this configuration to add nodes to which it should speak.
If we do not allow non-leader to do that, each node in the replica set will find itself isolated and will not be able to communicate with other nodes as it will not accept incoming configuration from the new primary as it would not be able to decide if there's a primary granted its knowledge of the configuration. This function acts as an oracle that gives knowledge of the other nodes IPs and then allows to reelect a leader in the replicaset.